### DISCUSSION TITLE
Spinner

### DISCUSSION THREAD

**Discussion by wartburggraaf** [nl-design-system/backlog#254] [Component suggestions]
![image](https://github.com/nl-design-system/backlog/assets/1439761/98a042d4-35ed-4d68-8722-145883971a04)

We willen graag meer leren over het **Spinner** component. In welke vorm, of vormen, wordt deze momenteel ingezet en toegepast binnen jouw organisatie?

Deel hieronder een link naar Figma, Storybook of screenshots. Met deze informatie kunnen we bewijzen dat het component onderdeel moet worden van de NL Design System component library.

Daarnaast wordt het delen van onderzoek gewaardeerd. Dus heb je zelf onderzoek gedaan en zijn er inzichten voor dit component? Dan horen we dit ook graag!

## Naam
Spinner

## Doel
Een animatie die aangeeft dat een proces nog bezig is, waardoor een deel van de pagina nog niet volledig is geladen of knoppen niet bedienbaar zijn.

## Gerelateerde componenten
[Progress Bar](https://github.com/orgs/nl-design-system/discussions/287)

## Onderzoek

[Loading Spinner - GOV.UK](https://github.com/alphagov/govuk-design-system-backlog/issues/28#issuecomment-465341005)
Discussion over de Loading Spinner. Er wordt gezegd dat uit gebruikersonderzoek naar reiswebsites bleek dat gebruikers baat hadden bij een minimum weergavetijd voor de Spinner. Hierdoor was er ook tijd voor zaken zoals het afvuren van aria-live events op een betekenisvolle manier.

[Loading & progress indicators — UI Components series - Medium](https://uxdesign.cc/loading-progress-indicators-ui-components-series-f4b1fc35339a)
Het artikel bespreekt het belang van laad- en voortgangsindicatoren die gebruikers informeren en betrekken tijdens wachttijden. Enkele interessante punten:
-  Er wordt aangegeven dat voor korte wachttijden (<1 seconde) geen indicatoren nodig zijn. Voor wachttijden van 1-3 seconden zijn simpele animaties, zoals een Spinner, geschikt. Bij wachttijden van 3-10 seconden zijn Progress Bars nuttig om de voortgang duidelijk te maken. 
- Er wordt aanbevolen om één globale voortgangsindicator weer te geven in plaats van meerdere indicatoren die de voortgang voor elke afzonderlijke Card bijvoorbeeld weergeven. Op deze manier kunnen gebruikers de algehele voortgang van het hele proces gemakkelijk begrijpen zonder overweldigd te raken.
- Door verschillende soorten laadindicatoren te combineren, ontstaan unieke en interessante ervaringen. Bijvoorbeeld door een Spinner te combineren met een Progress Bar, of animaties toe te voegen aan een Progress Circle.

[Skeleton Screens vs. Progress Bars vs. Spinners - Nielsen Norman Group](https://www.nngroup.com/videos/skeleton-screens-vs-progress-bars-vs-spinners/)
In de video worden verschillende soorten laadindicatoren besproken. Hieronder lees je wanneer je welke laadindicator het beste kunt gebruiken:
- Skeleton Screens zijn simpele visuele elementen die het lay-out van een pagina nabootsen en aangeven dat een pagina aan het laden is. Ze kunnen statisch of geanimeerd zijn en worden het best gebruikt voor volledige paginaladingen tussen de 2 en 10 seconden.
- Spinners lijken op Skeleton Screens, maar geven geen indicatie van de wachttijd. Ze zijn geïsoleerde elementen die een herhalende beweging maken en zijn ideaal voor het laden van individuele modules, zoals een video of een kaart op een dashboard. Spinners worden aanbevolen voor wachttijden van minder dan 10 seconden.
- Progress Bars tonen de voortgang van een taak, zoals een bestand dat wordt geüpload of gedownload. Ze zijn het best geschikt voor procesgerelateerde handelingen en worden sterk aanbevolen voor pagina’s die langer dan 10 seconden nodig hebben om te laden, omdat ze gebruikers laten zien hoe lang ze nog moeten wachten.

[Making an accessible loading button: Aria-disabled with friends - Bekk Christmas ](https://www.bekk.christmas/post/2023/24/accessible-loading-button)
In deze blog vertelt Bekk dat het ontwerpen van een toegankelijke Loading Button een echte hoofdbreker is. Er is geen eenvoudige one-size-fits-all-oplossing.

Rian geeft in [deze opmerking](https://github.com/orgs/nl-design-system/discussions/254#discussioncomment-10876733) aan dat een Spinner in een submit Button, waarbij de Button tekst daarna wordt aangepast, problematisch kan zijn voor de ondersteuning van screenreaders en het begrip van de gebruiker.

[Use “Delivery Date” Not “Shipping Speed” - Baymard Institute](https://baymard.com/blog/shipping-speed-vs-delivery-date)
Zie discussie [hieronder](https://github.com/orgs/nl-design-system/discussions/254#discussioncomment-9582047).

**Comment by wartburggraaf** [comment]
## DSO

De **Spinner** zien we terug in het component Progress Indicator

- [Als onderdeel van Modal in GitHub](https://github.com/dso-toolkit/dso-toolkit/blob/master/storybook/src/components/modal/modal.content.ts)
- [Documentatie website](https://storybook.dso-toolkit.nl/62.8.4/?path=/docs/html-css-progress-indicator--docs)
- [Storybook](https://storybook.dso-toolkit.nl/62.8.4/?path=/story/html-css-progress-indicator--medium)

---

### Screenshots / Varianten

**Storybook**

**Progress Indicator**
<img width="475" alt="Screenshot 2024-05-28 at 10 56 18" src="https://github.com/nl-design-system/backlog/assets/1439761/853b763c-3302-4dbf-a7ad-c9576b647e87">

---

### Notities

Toepassing binnen **Modal**
`<span class="dso-progress-indicator-spinner" role="progressbar" aria-labelledby="progress-indicator-label"></span>`

**Comment by wartburggraaf** [comment]
## DUO

[Documentatie website](https://uno.dfront.rijkscloud.nl/#/components/spinner)

---

### Screenshots / Varianten

**Documentatie website**

Naam variant: **Spinner**
Rationale variant: Een spinner is een visuele, geanimeerde indicator die aangeeft dat (een gedeelte van) een pagina nog bezig is met laden. Hoewel de spinner geen voortgang toont, laat het wel zien dat er iets gebeurt op de achtergrond. Dit voorkomt dat gebruikers te vroeg wegklikken.
<img width="163" alt="Screenshot 2024-05-28 at 11 00 15" src="https://github.com/nl-design-system/backlog/assets/1439761/66e5a22b-03d4-4d1d-8d90-be55f3430eb7">

---

### Notities

**Buttons** kunnen ook een **Spinner** bevatten als er met een klik op de knop veel data verwerkt moet worden. De button is tijdens het laden inactief.

**Comment by wartburggraaf** [comment]
## Open Formulieren

- [Storybook](https://open-formulieren.github.io/open-forms-sdk/?path=/story/pure-react-components-loader--loader)
- Figma (niet beschikbaar)

---

### Screenshots / Varianten

**Storybook**

Naam variant: **Loader**
Rationale variant: - 
<img width="80" alt="Screenshot 2024-05-28 at 11 13 00" src="https://github.com/nl-design-system/backlog/assets/1439761/af6788d3-d9d4-43bc-9474-526a7818c12f">

---

### Notities
`<span class="openforms-loading__spinner"></span>`

**Reply by sergei-maertens** [reply]
Wordt gebruikt in veel plaatsen:

* `main` element als er API calls zijn (bijvoorbeeld https://open-formulieren.github.io/open-forms-sdk/?path=/story/private-api-appointments-steps-5-confirmation--success)
* `<button aria-disabled="true">` als er gewacht wordt op netwerk verkeer

in React zouden we dit idealiter ook met `Suspense` kunnen gebruiken, maar zover zijn we nog niet

**Comment by wartburggraaf** [comment]
## Amsterdam

[Figma](https://www.figma.com/design/9IGm6IdPUYizBNGsUnueBd/%F0%9F%8E%A8-Amsterdam-Design-System?node-id=5562-45555&t=fB5OB8IFi0mQuKob-4)

---

### Screenshots / Varianten

![image](https://github.com/nl-design-system/backlog/assets/1439761/4e0b40d0-8cd1-4792-8659-4611d4de87e3)

---

### Notities

- Onder de naam: Spinner.

**Comment by wartburggraaf** [comment]
[Baymard](https://baymard.com/blog/shipping-speed-vs-delivery-date)

![research-media-file-0970b46fd564e2ca2858d58eb12c6a45](https://github.com/nl-design-system/backlog/assets/1439761/1ca5e432-48f3-4cd5-8f2c-419bfc03d2de)

Een redelijke oplossing is om aanvankelijk de (minder dan ideale) verzendsnelheid te tonen en vervolgens onmiddellijk bij het laden van de pagina een *AJAX-request* uit te voeren om de daadwerkelijke leveringsdata op te halen, de verzendsnelheden te vervangen door de datums zodra ze zijn geretourneerd.

Op deze manier worden de laadtijden van het betaalproces helemaal niet beïnvloed door de externe service: als de postservice API snel de geschatte leveringsdatums retourneert, worden de verzendsnelheden min of meer onmiddellijk vervangen, maar als de service toevallig uitvalt of traag is, heeft de gebruiker **nog steeds informatie** over de verzendopties (het is gewoon niet optimaal).

Als het gaat om het vertrouwen op een externe API om leveringsdatums te berekenen, is de kernafweging dat terwijl de feitelijke datumberekeningen worden vereenvoudigd (hoewel niet geëlimineerd), er *fallback*-oplossingen beschikbaar moeten zijn om te voorkomen dat de **snelheid en stabiliteit** van het afrekenen afhankelijk zijn van een extern systeem.

**Comment by Rozerinay** [comment]
## Utrecht

---

### Screenshots / Varianten

**Publieke website**

<img width="600" alt="image" src="https://github.com/user-attachments/assets/52edfd0d-0b4a-41c4-b77b-5883bd8a9f6f">

Pagina: [Melding maken](https://meldingen.utrecht.nl/incident/beschrijf)

---

### Notities
Spinner is heel kort zichtbaar voordat je in het formulier terechtkomt.

**Comment by Rozerinay** [comment]
## Tilburg

---

### Screenshots / Varianten

**Publieke website**

<img width="600" alt="image" src="https://github.com/user-attachments/assets/b57e43fd-0ee5-4254-a590-5902c77182bd">

Pagina: [Afspraak maken](https://winkel.tilburg.nl/formulieren/server/winkel-ui/IntakeAfspraakWaarvoor.html?sessionId=784ee03e-f32d-4e53-87d0-892711e8c1fb)

---

### Notities
Spinner is in combinatie met een bericht heel kort zichtbaar nadat je een Checkbox en andere input fields aanklikt.

**Comment by rianrietveld** [comment]
Naar aanleiding van de vraag van @Rozerinay wat overwegingen

> Tijdens de Developer Open Hour bespraken we de [Spinner](https://github.com/orgs/nl-design-system/discussions/254). Om deze Help Wanted te kunnen maken, hebben we meer praktijkvoorbeelden nodig. Yolijn gaf een goed voorbeeld van een loading state in een andere context: de Spinner in een Button.
> We zijn benieuwd naar de toegankelijkheid hiervan. Ik las in een artikel dat dit [vaak een uitdagend vraagstuk is](https://www.bekk.christmas/post/2023/24/accessible-loading-button). Weten onze a11y-experts hier meer over? Is de Spinner in een Button een voorbeeld dat we willen ondersteunen?
> 

Een spinner in een submit button en dan de naam wijzigen kan problematisch zijn voor screenreader support en begrip.
Zie https://adrianroselli.com/2020/12/be-careful-with-dynamic-accessible-names.html
Het gaat niet altijd goed bij de aankondiging voor screen readers.

Plaatsen van essentiele tekst info in de CSS wordt goed ondersteund voor assistive technology maar kan lastig worden als de pagina vertaald moet worden.

De andere oplossing, een tussenscherm met een spinner, heeft weer een ander probleem, zoals ik heb gezien bij Gravity Forms.
Deze is alleen zinvol als het wachten wat langer duurt en dat is afhankelijk van de snelheid van de submit. Screen reader feedback bij hele korte wachtijden voor een tussenpagina kan verwarrend zijn.

**Reply by MarjonBakker** [reply]
Roselli linkt in zijn artikel over dynamic accessible names naar zijn artikel over [Multi-Function Button](https://adrianroselli.com/2021/01/multi-function-button.html). Hij gebruikt daar een `span` met een `role=status` om de veranderingen van tekst in de button aan hulpsoftware te communiceren. Ik denk dat dat beter zou werken dan alleen op de buttontekst vertrouwen, eens?

**Reply by rianrietveld** [reply]
Zou iets zijn om uit te testen. De verandering wordt dan uitgesproken. De tekst moet dan wel contextueel kloppen, dat de announcement an sich betekenisvol is.
Voor de developers: Je kunt de role="status" niet rechtsreeks op de button zetten (want die heeft al de role van button), vandaar dat de extra span nodig is.

**Comment by Rozerinay** [comment]
## RVO

[Storybook](https://nl-design-system.github.io/rvo/?path=/docs/componenten-loader--documentatie)

---

### Screenshots / Varianten

<img width="960" height="260" alt="Scherm­afbeelding 2026-05-19 om 17 12 01" src="https://github.com/user-attachments/assets/7a7c8025-7995-48b9-8fbe-d798c151d196" />


---

### Notities

- Onder de naam: Loading spinner.
- Wordt gebruikt bij belangrijke 'verzend'-momenten.
- De gehele applicatie wordt tijdelijk geblokkeerd wanneer de Loading spinner zichtbaar is.
- Richtlijnen beschikbaar.

**Comment by Rozerinay** [comment]
## ✨ Dit component is nu Help Wanted ✨

Help je mee hem door de Community stappen te krijgen?
[Spinner op het Community bord](https://github.com/orgs/nl-design-system/projects/29/views/1?filterQuery=spinner)

**Comment by Robbert** [comment]
Bij Terraform ben ik 'm tegengekomen in een Status Badge en in een Heading Group.

- Heading Group <img width="335" alt="Screenshot 2024-10-30 at 16 26 18" src="https://github.com/user-attachments/assets/0750d7e3-96d6-4aa5-af6e-f03c855e289d">
- Status Badge <img width="123" alt="Screenshot 2024-10-30 at 16 25 50" src="https://github.com/user-attachments/assets/8fa5f634-31d0-4a26-9666-059c103e40ef">

**Comment by jeffreylauwers** [comment]
Bij CBIG los tegengekomen. Maar ook in combinatie met tekst, een button en een input field:
https://designsystem.cibg.nl/componenten/lader/
